File: src/main/java/org/janelia/saalfeldlab/util/n5/metadata/N5PainteraDataMultiScaleGroup.java
package org.janelia.saalfeldlab.util.n5.metadata;

import net.imglib2.realtransform.AffineGet;
import net.imglib2.realtransform.AffineTransform3D;
import org.janelia.saalfeldlab.n5.universe.metadata.N5SpatialDatasetMetadata;
import org.janelia.saalfeldlab.n5.universe.metadata.SpatialMultiscaleMetadata;

import java.util.function.Predicate;
import java.util.regex.Pattern;

public abstract class N5PainteraDataMultiScaleGroup extends SpatialMultiscaleMetadata<N5SpatialDatasetMetadata> {

	public static final Predicate<String> SCALE_LEVEL_PREDICATE = Pattern.compile("^s\\d+$").asPredicate();
	protected final N5PainteraDataMultiScaleMetadata dataGroup;

	public N5PainteraDataMultiScaleGroup(String basePath, final N5PainteraDataMultiScaleMetadata dataGroup) {

		super(basePath, dataGroup.getChildrenMetadata());
		this.dataGroup = dataGroup;
	}

	public N5PainteraDataMultiScaleMetadata getDataGroupMetadata() {

		return dataGroup;
	}

	@Override public AffineGet spatialTransform() {

		return dataGroup.spatialTransform();
	}

	@Override public String unit() {

		return dataGroup.unit();
	}

	@Override public AffineTransform3D spatialTransform3d() {

		return dataGroup.spatialTransform3d();
	}

	@Override
	public AffineGet[] spatialTransforms() {
		return dataGroup.spatialTransforms3d();
	}

	@Override
	public AffineTransform3D[] spatialTransforms3d() {
		return dataGroup.spatialTransforms3d();
	}
}

File: src/main/kotlin/org/janelia/saalfeldlab/util/n5/metadata/N5PainteraDataMultiscaleGroup.kt
package org.janelia.saalfeldlab.util.n5.metadata

import net.imglib2.realtransform.AffineGet
import net.imglib2.realtransform.AffineTransform3D
import org.janelia.saalfeldlab.n5.universe.metadata.N5SpatialDatasetMetadata
import org.janelia.saalfeldlab.n5.universe.metadata.SpatialMultiscaleMetadata
import org.janelia.saalfeldlab.n5.universe.metadata.axes.Axis
import org.janelia.saalfeldlab.n5.universe.metadata.axes.AxisMetadata

/**
 * A Paintera group whose scale levels live in a child `data` group; the spatial transforms and axes all delegate
 * to that [dataGroupMetadata].
 */
abstract class N5PainteraDataMultiscaleGroup(
	basePath: String,
	val dataGroupMetadata: N5PainteraDataMultiscaleMetadata
) : SpatialMultiscaleMetadata<N5SpatialDatasetMetadata>(basePath, dataGroupMetadata.childrenMetadata), AxisMetadata {

	override fun spatialTransform(): AffineGet = dataGroupMetadata.spatialTransform()

	override fun unit(): String = dataGroupMetadata.unit()

	override fun spatialTransform3d(): AffineTransform3D = dataGroupMetadata.spatialTransform3d()

	override fun spatialTransforms(): Array<out AffineGet> = dataGroupMetadata.spatialTransforms3d()

	override fun spatialTransforms3d(): Array<AffineTransform3D> = dataGroupMetadata.spatialTransforms3d()

	override fun getAxes(): Array<Axis> = dataGroupMetadata.childrenMetadata[0].axes
}
